runtime/vm/jniinv.c

```diff
@@ -346,6 +349,11 @@ protectedDestroyJavaVM(J9PortLibrary* portLibrary, void * userData)
 		omrthread_monitor_exit(vm->runtimeFlagsMutex);
 	}
 
+#ifdef J9ZTPF
+	/* run z/TPF OS exit hook */
+	cjvm_jvm_shutdown_hook();
```

Comment:
This seems too early, the sidecarShutdown() below is going to run java code, i.e. java.lang.Shutdown.shutdown().

Reply:
Good point let me revisit the placement and then re-push the commit.

Additionally, I need to verify that it can only ever be executed exactly once.  I can't tell if a race condition can occur in protectedDestroyJavaVM().

I notice later on in the same function there is a check for vm->exitHook in two different return scenarios... maybe the z/TPF shutdown hook should be positioned in the same locations?

Reply:
> maybe the z/TPF shutdown hook should be positioned in the same locations?

Ya, makes sense.